### PR TITLE
feat: add support for run task results callback

### PR DESCRIPTION
**Pull request by karvounis-form3**
## Description

This PR aims to address the missing support for Run Tasks Integration API. I have already raised an issue https://github.com/hashicorp/go-tfe/issues/862 about this.
It introduces the following:
- RunTaskRequest struct which can be used to objectify the request you get from a TFC/E run task
- RunTasksCallback interface which can be used to update the run task with status and messages
  - TaskResultsCallbackOptions defines the attributes required in order to perform the update

## Testing plan

Unfortunately, it is not easy to test the Update function because you first need a running server which listens for incoming TFC/E requests. Then, you need a workspace run which is configured to use that run task endpoint.

However, I added the following unit tests and their output when run locally:

- TestTaskResultsCallbackRequestOptions_Validate

Screenshot 2024-06-19 at 8 27 28 AM

- TestTaskResultsCallbackRequestOptions_Marshal

Screenshot 2024-06-19 at 8 28 01 AM

- TestTaskResultsCallbackUpdate_Validate

Screenshot 2024-06-19 at 8 28 25 AM

- TestTaskResultsCallbackUpdate

Screenshot 2024-06-19 at 8 28 52 AM

## External links

https://github.com/hashicorp/terraform-run-task-scaffolding-go this HashiCorp repo contains part of the code that can be found here.

### CODE REVIEW

File: run_task_request.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+package tfe
+
+import (
+	"time"
+)
+
+const (
+	// RunTaskRegistrationRunId is the string that TFC/E sends when it just wants to register the Run Task.
+	RunTaskRegistrationRunId = "run-xxxxxxxxxxxxxxxx"
+)
```
Comment (sebasslash) [inline run_task_request.go:13]: Would like to understand this more, along with the `IsEndpointValidation` check. Is the intent to ensure `r.RunID` is a valid run external ID (in the format of `RunTaskRegistrationRunID`)? As I read it, `IsEndpointValidation` will only return true if `r.RunID == "run-xxxxxxxxxxxx"` which should never be the case. 

If we do want to ensure that it's in proper external ID format, use the helper [validStringID()](https://github.com/hashicorp/go-tfe/blob/193d7a6b5066d8f4528ae0a38da39e7c82360761/validations.go#L29)

**Review by sebasslash — CHANGES_REQUESTED** [review]
Awesome, thanks for submitting this PR & sorry for the delay in reviewing. Some things to chew on in the meantime ⬇️

File: run_task_results_callback.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	// Message A short message describing the status of the task.
+	Message string `json:"message,omitempty"`
+	// URL that the user can use to get more information from the external service
+	URL string `json:"url,omitempty"`
+}
+
+type TaskResultsCallbackRelationships struct {
+	// Outcomes A run task result may optionally contain one or more detailed outcomes, which improves result visibility and content in the Terraform Cloud user interface.
+	// https://developer.hashicorp.com/terraform/enterprise/api-docs/run-tasks/run-tasks-integration#outcomes-payload-body
+	Outcomes *TaskResultsCallbackRelationshipsOutcomes `json:"outcomes"`
```
[the remark this answers is not in the document]
Comment (karvounis-form3) [reply run_task_results_callback.go:73]: Hey @sebasslash thanks for the comment! I will work on that asap!

[the remark this answers is not in the document]
Comment (karvounis-form3) [reply run_task_results_callback.go]: I made the changes and I have confirmed that they work! Thanks for the amazing suggestion. I learned something new!

File: run_task_results_callback.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// TaskResultOutcome represents a detailed TFC/E run task outcome, which improves result visibility and content in the TFC/E UI.
+// https://developer.hashicorp.com/terraform/enterprise/api-docs/run-tasks/run-tasks-integration#outcomes-payload-body
+type TaskResultOutcome struct {
+	Type        string                      `jsonapi:"primary,task-result-outcomes"`
+	OutcomeID   string                      `jsonapi:"attr,outcome-id,omitempty"`
+	Description string                      `jsonapi:"attr,description,omitempty"`
+	Body        string                      `jsonapi:"attr,body,omitempty"`
+	URL         string                      `jsonapi:"attr,url,omitempty"`
+	Tags        map[string][]*TaskResultTag `jsonapi:"attr,tags,omitempty"`
+}
```
Comment (sebasslash) [inline run_task_results_callback.go:65]: Move these above the `Update()` call for visibility. The style ordering for a file is:
1. Interface definition
2. Struct definitions
3. Receiver method implementations.

**Review by sebasslash — CHANGES_REQUESTED** [review]
So the implementation look great ✅ However I do think we ought to add some unit tests around this behavior, particularly ensuring that things are being serialized correctly (`TaskResultTag` gives me some pause given mix-&-matching of `json/jsonapi` struct tags).

Writing a true integration test would not be plausible here, as it would require the test function to expose a publicly available server. 

What sort of unit tests might be possible here?
- Test for validation errors, scenarios that would result in `ErrInvalidTaskResultsCallbackStatus`, `ErrInvalidCallbackURL`, `ErrInvalidAccessToken` and `ErrInvalidTaskResultsCallbackType`. 
- Ensure the `Authorization` header is set to the correct token provided in the request sent from TFC. In the test function, you can spin up an [httptest.Server](https://pkg.go.dev/net/http/httptest#example-Server), define an endpoint that validates the header is the correct token, and set that endpoint URL to `TaskResultCallbackURL` in a mocked `RunTaskRequest`. 
- Ensure `TaskResultTag` is serialized correctly.

Last thing, make sure you define the server in `helper_test.go`. An example signature can be: 
```
func runTaskCallbackMockServer(t *testing.T) *httptest.Server {}
```

Once you've added tests, make sure to run them locally and add the test output to the PR description. CI won't fire given this is a fork's branch, so once everything is a-ok I will rebase this PR onto a local branch to trigger CI.

File: run_task_results_callback.go (hunk 3 above)
Comment (karvounis-form3) [reply run_task_results_callback.go:44]: I moved them up just before the `Update` function

**Comment by karvounis-form3** [comment]
@sebasslash thank you for your suggestions to improve the code. I added the unit tests and added screenshots from running the tests locally in the PR body as requested. Can you please review the PR once more?

**Comment by karvounis-form3** [comment]
Fixed code based on `golangci-lint` [suggestions](https://github.com/hashicorp/go-tfe/actions/runs/9576670518/job/26657617891?pr=867)

**Comment by karvounis-form3** [comment]
@Maed223 maybe you can rerun the CI?

**Comment by karvounis-form3** [comment]
@Maed223 https://github.com/hashicorp/go-tfe/actions/runs/9664123031/job/26672604714#step:2:1 I do not understand what actions to take to fix that issue

**Comment by karvounis-form3** [comment]
@sebasslash @Maed223 any thoughts on this?

**Review by sebasslash — APPROVED** [review]

**Comment by sebasslash** [comment]
Since this PR comes from a fork, I will rebase your changes onto a local branch so that your changes execute against CI.

**Comment by karvounis-form3** [comment]
@sebasslash can we please prioritise merging that PR? We need that code

**Comment by sebasslash** [comment]
Closing in favor of #929

**Pull request closed without merging**